core/src/serde/de.rs

```diff
@@ -0,0 +1,693 @@
+            }
+
+            #[inline]
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::String(String::from(value)))
+            }
+
+            #[inline]
+            fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                self.visit_byte_buf(v.to_owned())
+            }
+
+            #[inline]
+            fn visit_byte_buf<E>(self, v: Vec<u8>) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Bytes(v))
+            }
+
+            #[inline]
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Integer(v.into()))
+            }
+
+            #[inline]
+            fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Integer(v.into()))
+            }
+
+            #[inline]
+            fn visit_i128<E>(self, v: i128) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Integer(v))
+            }
+
+            #[inline]
+            fn visit_f64<E>(self, v: f64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Float(v))
+            }
+
+            #[inline]
+            fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Bool(v))
+            }
+
+            #[inline]
+            fn visit_none<E>(self) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                Ok(Ipld::Null)
+            }
+
+            #[inline]
+            fn visit_seq<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::SeqAccess<'de>,
+            {
+                let mut vec = Vec::new();
+
+                while let Some(elem) = visitor.next_element()? {
+                    vec.push(elem);
+                }
+
+                Ok(Ipld::List(vec))
+            }
+
+            #[inline]
+            fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::MapAccess<'de>,
+            {
+                let mut values = BTreeMap::new();
+
+                while let Some((key, value)) = visitor.next_entry()? {
+                    values.insert(key, value);
+                }
+
+                Ok(Ipld::Map(values))
+            }
+
+            /// Newtype structs are only used to deserialize CIDs.
+            #[inline]
+            fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
```

Comment:
I still feel like I'm missing something. Are we talking about a call like this:

```rust
Ipld::deserialize(something_that_impls_Deserializer);
```

You'd want to deserialize from some codec into `Ipld`. This would usually be some blob that implements a DAG codec. So now your question is: "could we get any strange results, when some random `Deserializer` is used?". I don't know why you would do that, but let's say you do.

Your `Deserializer` implementation would need to have a code path that calls `visit_newtype_struct`. Then that `Deserializer` also needs to have a `deserialize_bytes` implementation. That could of course call any `visit_*` it likes. But luckily we pass in the [`BytesToCidVisitor`](https://github.com/multiformats/rust-cid/blob/6cf0f6768e5b9e12f802e23bb63673d729898eef/src/serde.rs#L38-L68), which only implements `visit_bytes` and `visit_seq`. So all other `visit_*` calls would error, no harm here. When it calls `visit_bytes` or `visit_seq` we get some bytes or a sequence. If those form by coincidence a valid CID, we would return that CID. Else we would again error.

To conclude: a random `Deserializer` implementation is likely to error. Only if it leads to a byte sequence that leads to a valid CID, a CID is returned. And I would say if it does, it was likely a codec that was crafted to support CIDs.

Reply:
I'm setting the bar to "we can prove to ourselves that this is correct" not "we can't think of a way for this to go wrong".